src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreCycleDetector.cs

```diff
@@ -48,9 +48,21 @@ internal sealed class PackageRestoreCycleDetector(
     private int _nuGetRestoreSuccesses;
     private int _nuGetRestoreCyclesDetected;
     private int _counter;
+    private ProjectConfiguration? _priorActiveProjectConfiguration;
 
-    public async Task<bool> IsCycleDetectedAsync(Hash hash, CancellationToken cancellationToken)
+    public async Task<bool> IsCycleDetectedAsync(Hash hash, ProjectConfiguration activeProjectConfiguration, CancellationToken cancellationToken)
     {
+        _priorActiveProjectConfiguration ??= activeProjectConfiguration;
+
+        if (!Equals(_priorActiveProjectConfiguration, activeProjectConfiguration))
+        {
+            // The active configuration changed. Treat this as a reset. This prevents flagging a
+            // cycle when the user switches configurations repeatedly in certain repositories.
+            _priorActiveProjectConfiguration = activeProjectConfiguration;
+            Reset();
+            return false;
+        }
```

Comment:
This is the core fix. If the active configuration changes, clear all state in our detection logic. Most of the other changes in this PR exist to support this code.